x-pack/plugins/apm/public/components/shared/redirect_with_offset/index.tsx

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { useLocation, Redirect } from 'react-router-dom';
+import qs from 'query-string';
+import React from 'react';
+import moment from 'moment';
+import { useApmRouter } from '../../../hooks/use_apm_router';
+import { getDateRange } from '../../../context/url_params_context/helpers';
+
+import {
+  TimeRangeComparisonEnum,
+  getSelectOptions,
+} from '../../../components/shared/time_comparison/get_comparison_options';
+
+export function RedirectWithOffset({
+  children,
+}: {
+  children: React.ReactElement;
+}) {
+  const location = useLocation();
+  const query = qs.parse(location.search);
+
+  const apmRouter = useApmRouter();
+  const matchingRoutes = apmRouter.getRoutesToMatch(location.pathname);
+  const matchesRoute = matchingRoutes.some((route) => {
+    return (
+      route.path === '/services' ||
+      route.path === '/traces' ||
+      route.path === '/service-map' ||
+      route.path === '/backends' ||
+      route.path === '/services/{serviceName}' ||
+      route.path === '/service-groups' ||
+      location.pathname === '/' ||
+      location.pathname === ''
+    );
+  });
+
+  if (
+    'comparisonType' in query &&
+    'rangeFrom' in query &&
+    'rangeTo' in query &&
+    matchesRoute &&
+    Object.values(TimeRangeComparisonEnum).includes(
+      query.comparisonType as TimeRangeComparisonEnum
+    )
+  ) {
+    const { rangeFrom, rangeTo, comparisonType, ...queryRest } = query;
+
+    const { start, end } = getDateRange({
+      rangeFrom: rangeFrom as string,
+      rangeTo: rangeTo as string,
+    });
+
+    const momentStart = moment(start);
+    const momentEnd = moment(end);
+
+    const offset = getSelectOptions({
```

Comment:
nit: perhaps the relevant part should be abstracted out of the getSelectOptions function, calling this here seems kind of odd.

Reply:
Discussed another option in https://github.com/elastic/kibana/pull/127812#discussion_r833147083